### PR TITLE
feat: add platform option to push command

### PR DESCRIPTION
**Pull request by TerryHowe**
**What this PR does / why we need it**:

Add the platform option to the push command for example:
```
oras push --plain-http localhost:15000/oras:darwin,arm64  --artifact-platform darwin/arm64 --artifact-type 'application/vnd.oci.image.config.v1+json'  bin/darwin/arm64/oras:application/x-elf
```

Create the manifest for example:
```
oras manifest index create localhost:15000/oras:v1 sha256:ed3691788db69623790c46256dd11d8d6edc31be6e4762b53c74dfc3bb6792cf
```

The results
```
% oras manifest fetch --pretty localhost:15000/oras@sha256:c9e54542075ad502900d11b9b5d5bdb4b974ad580842515684d3fbb4849e2bb1
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "application/vnd.oci.image.config.v1+json",
  "config": {
    "mediaType": "application/vnd.unknown.config.v1+json",
    "digest": "sha256:e9302bbb2fb8f6c2df866d3c4e41917849442f89a575f36f43366a7279804f70",
    "size": 38
  },
  "layers": [
    {
      "mediaType": "application/x-elf",
      "digest": "sha256:69c16b8386b7949fe37ff2356296b6331c1e6d5fbbc3e616192cc10ece55ea8d",
      "size": 11673840,
      "annotations": {
        "org.opencontainers.image.title": "bin/darwin/amd64/oras"
      }
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-09-16T16:54:48Z"
  }
}
% oras manifest fetch-config --pretty localhost:15000/oras@sha256:c9e54542075ad502900d11b9b5d5bdb4b974ad580842515684d3fbb4849e2bb1
{
  "architecture": "amd64",
  "os": "darwin"
}
```
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Closes https://github.com/oras-project/oras/issues/1066
Partial https://github.com/oras-project/oras/issues/1053

**Please check the following list**:
- [ ]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [x]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [x]  Do all new files have an appropriate license header?

### CODE REVIEW

File: cmd/oras/root/push.go
Comment (qweeah) [inline cmd/oras/root/push.go:105]: CC @FeynmanZhou for the flag naming. `attach` already using it so better call the new one `--config-platform`

File: cmd/oras/root/push.go
Comment (qweeah) [inline cmd/oras/root/push.go:47]: Should add this to `Packer` options

File: cmd/oras/root/push.go
Comment (qweeah) [inline cmd/oras/root/push.go:165]: Use `CheckMutuallyExclusiveFlags` instead. The output of this function is hard to read.

File: cmd/oras/root/push.go
Comment (wangxiaoxuan273) [inline cmd/oras/root/push.go:206]: nit: This part can be replaced by `func NewDescriptorFromBytes(mediaType string, content []byte) ocispec.Descriptor` of oras-go. Reference: [NewDescriptorFromBytes](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0/content#NewDescriptorFromBytes)

File: cmd/oras/root/push.go
Comment (wangxiaoxuan273) [reply cmd/oras/root/push.go:47]: But if this is added to `Packer`, we need to change `oras attach` since it already has a `--platform` flag.

Comment (qweeah) [reply cmd/oras/root/push.go:47]: 1. Both `attach` and `push` should support customized platform. 
2. It's inconsistent if we use `--platform` for customizing config platform in `push`, while `--platform` for selecting subject platform in `attach`

File: cmd/oras/root/push.go
Comment (shizhMSFT) [reply cmd/oras/root/push.go:105]: It is a good point that `--platform` means differently in `oras push` and `oras attach`. We need a better flag name.

Comment (shizhMSFT) [reply cmd/oras/root/push.go:105]: `--config-platform` does not sound like a good name. Would it be better to name it `--artifact-platform`? That will be consistent with `--artifact-type`.

File: cmd/oras/root/push.go
Comment (shizhMSFT) [inline cmd/oras/root/push.go:148]: They can be used together, can't they?

File: cmd/oras/root/push.go
Comment (shizhMSFT) [inline cmd/oras/root/push.go:203]: This value depends on the image-spec version.

File: cmd/oras/root/push.go
Comment (Wwwsylvia) [reply cmd/oras/root/push.go:148]: Indeed they can be used together. We can populate the `artifactType` as `mediaType` into the config descriptor.

File: cmd/oras/root/push.go
Comment (Wwwsylvia) [reply cmd/oras/root/push.go:203]: If the `image-spec` version is `v1.0`. The media type can also be the value of `--artifact-type`.

File: cmd/oras/root/push.go
Comment (TerryHowe) [reply cmd/oras/root/push.go]: Done.

File: cmd/oras/root/push.go
Comment (TerryHowe) [reply cmd/oras/root/push.go]: Done.

File: cmd/oras/root/push.go
Comment (TerryHowe) [reply cmd/oras/root/push.go]: Done.

File: cmd/oras/root/push.go
Comment (qweeah) [reply cmd/oras/root/push.go]: Let not merge it to `Packer` since this PR is all about `push`

File: cmd/oras/root/push.go
Comment (TerryHowe) [reply cmd/oras/root/push.go]: Done

**Review by Wwwsylvia — APPROVED** [review]
LGTM with nit, but I am not a maintainer.

File: cmd/oras/root/push.go
Comment (TerryHowe) [reply cmd/oras/root/push.go]: Done

**Review by sabre1041 — APPROVED** [review]
LGTM

File: cmd/oras/root/push.go
Comment (TerryHowe) [reply cmd/oras/root/push.go]: This does mean pull will use `--platform` and push `--artifact-platform` though which does seem a bit inconsistent to me.

Comment (qweeah) [reply cmd/oras/root/push.go]: It's hard to choose a good flag name.

Some historical view on this: `--platform` was firstly introduced to `oras manifest to select an image with specific platform from a multi-arch image. Then the same UX was applied to `oras discover` and `oras attach` for selecting a platform-specific subject artifact.

File: cmd/oras/internal/option/platform.go
Comment (qweeah) [inline cmd/oras/internal/option/platform.go:86]:
```suggestion
	opts.FlagDescription = "set artifact platform"
```

File: cmd/oras/root/push.go
Comment (qweeah) [inline cmd/oras/root/push.go:168]: This can be done in `PreRunE` before `option.Parse`

File: cmd/oras/root/push.go
Comment (qweeah) [inline cmd/oras/root/push.go:125]: This block handles `oras push --config xxx --artifact-type xxx ...`: 

- For v1.1 artifacts, this usage invalid. 
- Otherwise best-effort try to pack v1.0 artifacts.


Since `--artifact-platform` cannot be used with `--config`, there is no need to change this code block.

File: cmd/oras/internal/option/platform.go
Comment (qweeah) [inline cmd/oras/internal/option/platform.go:87]:
```suggestion
	opts.FlagDescription = "set artifact platform"
	opts.Platform.AppFlags(fs)
```

**Review by qweeah — CHANGES_REQUESTED** [review]

File: cmd/oras/root/push.go
Comment (qweeah) [reply cmd/oras/root/push.go:125]: Only need to make sure: `--artifact-platform` is not used for v1.0 artifact in the below `opts.PackVersion` switching block.

Comment (TerryHowe) [reply cmd/oras/root/push.go:125]: Made the change I think you requested. I do need to think about this some more though.

File: cmd/oras/root/push.go
Comment (TerryHowe) [inline cmd/oras/root/push.go:151]: I'm not sure what the v1.0 rule should be.

Comment (qweeah) [reply cmd/oras/root/push.go:151]: `--artifact-platform` specifies content of config blob, `--artifact-type` specifies `mediaType` of config blob, so both flags can be used for OCI v1.0 image.

File: cmd/oras/root/push.go
Comment (qweeah) [reply cmd/oras/root/push.go:125]: > Only need to make sure: --artifact-platform is not used for v1.0 artifact in the below opts.PackVersion switching block.

This is wrong actually. Sorry for the back-and-forth change.

**Review by qweeah — CHANGES_REQUESTED** [review]

File: cmd/oras/root/push.go
Comment (qweeah) [reply cmd/oras/root/push.go:125]: This is need if we are going to allow v1.1 `application/vnd.unknown.artifact.v1` artifact to have platform information in config blob.

Comment (qweeah) [reply cmd/oras/root/push.go:125]: @shizhMSFT If `--platform` is used without `--artifact-type` and `image-spec` is not provided, should we pack v1.1 or v1.0 image?

Comment (qweeah) [reply cmd/oras/root/push.go:125]: Currently if `--config` is not provided and `image-spec` is not provided, `oras` will pack v1.1 artifact
```console
➜  ~ oras push localhost:5000/test:type
✓ Uploaded  application/vnd.oci.empty.v1+json                                                                   2/2  B 100.00%   34ms
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                      535/535  B 100.00%     0s
  └─ sha256:5f56fc622e5f67d225a27be2c3637a606880c23aa98a2561b3797c4b54ac9fd5
Pushed [registry] localhost:5000/test:type
ArtifactType: application/vnd.unknown.artifact.v1
Digest: sha256:5f56fc622e5f67d225a27be2c3637a606880c23aa98a2561b3797c4b54ac9fd5
➜  ~ oras manifest fetch --pretty localhost:5000/test@sha256:5f56fc622e5f67d225a27be2c3637a606880c23aa98a2561b3797c4b54ac9fd5
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "application/vnd.unknown.artifact.v1",
  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json",
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2,
    "data": "e30="
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.empty.v1+json",
      "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
      "size": 2,
      "data": "e30="
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-10-10T06:29:05Z"
  }
}
```

Comment (qweeah) [reply cmd/oras/root/push.go:125]: If `--config` is used with `--artifact-type`, `oras` will also pack v1.1 artifact
```console
➜  ~ oras push localhost:5000/test:v1.1 --artifact-type test/123 --config 123
✓ Exists    application/vnd.oci.empty.v1+json                                                                   2/2  B 100.00%     0s
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded  application/vnd.unknown.config.v1+json                                                              3/3  B 100.00%   39ms
  └─ sha256:ca3d163bab055381827226140568f3bef7eaac187cebd76878e0b63e9e442356
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                      499/499  B 100.00%     0s
  └─ sha256:d80e321ea40ab36b661693ddf507f425801819481949d6bdbb6ec52f503139ae
Pushed [registry] localhost:5000/test:v1.1
ArtifactType: test/123
Digest: sha256:d80e321ea40ab36b661693ddf507f425801819481949d6bdbb6ec52f503139ae
➜  ~ oras manifest fetch --pretty localhost:5000/test@sha256:d80e321ea40ab36b661693ddf507f425801819481949d6bdbb6ec52f503139ae
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "test/123",
  "config": {
    "mediaType": "application/vnd.unknown.config.v1+json",
    "digest": "sha256:ca3d163bab055381827226140568f3bef7eaac187cebd76878e0b63e9e442356",
    "size": 3
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.empty.v1+json",
      "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
      "size": 2,
      "data": "e30="
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-10-10T06:34:43Z"
  }
}
```

If config media type is not provided, it will be set to `application/vnd.unknown.config.v1+json`. We can follow this practice for `--platform` and bake v1.1 artifact for `oras push --platform xxx --artifact xxx` usage.

Comment (shizhMSFT) [reply cmd/oras/root/push.go:125]: > @shizhMSFT If `--platform` is used without `--artifact-type` and `image-spec` is not provided, should we pack v1.1 or v1.0 artifact? (I would prefer v1.1)

Let's follow the default value of `image-spec` when this flag is not provided.

Comment (qweeah) [reply cmd/oras/root/push.go:125]: We make `oras push --config xxx` craft v1.0 artifact to avoid breaking change. `oras push --platform xxx` should create v1.1 artifact. The suggested change should be applied.

File: cmd/oras/root/push.go
Comment (qweeah) [reply cmd/oras/root/push.go:151]:
```suggestion
```

Comment (TerryHowe) [reply cmd/oras/root/push.go:151]: Since the code above this does not allow specifying `--artifact-type` and `--config` at the same time I thought it made sense to not allow `--artifact-typ` and `--artifact-platform` since platform is just a shortcut for config.

Comment (qweeah) [reply cmd/oras/root/push.go:151]: We do not allow specifying `--artifact-type` and `--config` at the same time because **those flags function are overlapped, user might specify different media types of config descriptor**, like `--artifact-type aaa --config xxx:bbb`, `aaa` and `bbb` conflicts.

But there is no such potential conflict between `--artifact-type` and `--artifact-platform`

File: cmd/oras/root/push.go
Comment (TerryHowe) [reply cmd/oras/root/push.go:128]: v1.1 image spec is default except for that case you mention `--config` and no `--artifact-type`

https://github.com/oras-project/oras/blob/main/cmd/oras/root/push.go#L122-L128

Comment (qweeah) [reply cmd/oras/root/push.go:128]: Yes, and we should make v1.1 default when `--artifact-platform xxx` is used regardless of `--artifact-type`.

Having `|| opts.Platform.Platform != nil` here will make v1.0 default when `--artifact-platform` used without `--artifact-type`.

Comment (TerryHowe) [reply cmd/oras/root/push.go]: Done

File: cmd/oras/root/push.go
Comment (qweeah) [inline cmd/oras/root/push.go:204]:
```suggestion
		mediaType := opts.artifactType
		if mediaType == "" {
			mediaType = opts.artifactType
		}
		desc := content.NewDescriptorFromBytes(mediaType, blob)
```

**Review by qweeah — APPROVED** [review]
LGTM with minor suggestion

File: cmd/oras/root/push.go
Comment (TerryHowe) [reply cmd/oras/root/push.go]: Manually made this change because oras.MediaTypeUnknownConfig.

Comment (TerryHowe) [reply cmd/oras/root/push.go]: With this change, the e2e test fails because `oras push --oci-layout /tmp/ginkgo1771247066:e2e --artifact-platform darwin/arm64 foobar/bar -v`

Results in:
```
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "application/vnd.unknown.artifact.v1",
  "config": {
    "mediaType": "application/vnd.unknown.artifact.v1",
    "digest": "sha256:e94c0ba80a1157ffab5b5c6656fffc089c6446c7ed0604f3382910d1ef7dd40d",
    "size": 38
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar",
      "digest": "sha256:d1bc8d3ba4afc7e109612cb73acbdddac052c93025aa1f82942edabb7deb82a1",
      "size": 5,
      "annotations": {
        "org.opencontainers.image.title": "foobar/bar"
      }
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-10-15T16:14:58Z"
  }
}
```

instead of 
```
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "artifactType": "application/vnd.unknown.artifact.v1",
  "config": {
    "mediaType": "application/vnd.unknown.config.v1+json",
    "digest": "sha256:e94c0ba80a1157ffab5b5c6656fffc089c6446c7ed0604f3382910d1ef7dd40d",
    "size": 38
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar",
      "digest": "sha256:d1bc8d3ba4afc7e109612cb73acbdddac052c93025aa1f82942edabb7deb82a1",
      "size": 5,
      "annotations": {
        "org.opencontainers.image.title": "foobar/bar"
      }
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2024-10-15T16:20:21Z"
  }
}
```

Comment (qweeah) [reply cmd/oras/root/push.go:204]: I think the former is preferred since user doesn't specify `--artifact-spec v1.0` or `--config`?

Comment (TerryHowe) [reply cmd/oras/root/push.go:204]: This is a problem tho 
```
  "artifactType": "application/vnd.unknown.artifact.v1",
  "config": {
    "mediaType": "application/vnd.unknown.artifact.v1",
```

**Comment by qweeah** [comment]
Sorry merging https://github.com/oras-project/oras/pull/1500/commits/55cf4260c98bed4c34b3cd612e44d0d0867774f5 is not intended, I was playing with the PR locally and the upstream is accidentally set to this PR.

**Comment by TerryHowe** [comment]
> Sorry merging [55cf426](https://github.com/oras-project/oras/commit/55cf4260c98bed4c34b3cd612e44d0d0867774f5) is not intended, I was playing with the PR locally and the upstream is accidentally set to this PR.

reverted

File: cmd/oras/root/push.go
Comment (qweeah) [reply cmd/oras/root/push.go:204]: I think switch-casing on the default value is a better option than setting the value.
```suggestion
                mediaType := opts.artifactType
		if mediaType == "" {
			switch opts.PackVersion {
			case oras.PackManifestVersion1_0:
				mediaType = ocispec.MediaTypeImageConfig
			case oras.PackManifestVersion1_1:
				mediaType = oras.MediaTypeUnknownArtifact
			}
		}
		desc := content.NewDescriptorFromBytes(mediaType, blob)
```

Comment (qweeah) [reply cmd/oras/root/push.go:204]: > This is a problem tho
> 
> ```
>   "artifactType": "application/vnd.unknown.artifact.v1",
>   "config": {
>     "mediaType": "application/vnd.unknown.artifact.v1",
> ```

This is valid for v1.1 though.

Comment (TerryHowe) [reply cmd/oras/root/push.go:204]: It may be valid, but I don't think that was the intention for the user

Comment (qweeah) [reply cmd/oras/root/push.go:204]: User doesn't specify `--artifact-type` so we cannot know what `config.mediaType` user is expecting.

If user what to specify `artifactType` and `config.mediaType` for v1.1 at the same time, they should use `--artifact-type <artifactType> --config <path>:<config.mediaType>`.

Comment (qweeah) [reply cmd/oras/root/push.go]:
```
  "artifactType": "application/vnd.unknown.artifact.v1",
  "config": {
    "mediaType": "application/vnd.unknown.artifact.v1",
```
This LGTM and it is compliant to [v1.1 artifact with config](https://oras.land/docs/concepts/artifact#artifacts-with-config)

Comment (TerryHowe) [reply cmd/oras/root/push.go]: The reason the test coverage failed was because https://github.com/oras-project/oras/blob/main/cmd/oras/root/push.go#L144 means `opts.artifactType` is never empty here, so I got rid of all the unnecessary logic.

File: cmd/oras/root/push.go
Comment (shizhMSFT) [inline cmd/oras/root/push.go:203]: This logic works for OCI spec v1.0. However, it does not work very well for OCI spec v1.1. And I agree with you w.r.t. https://github.com/oras-project/oras/pull/1500#discussion_r1801538753

Comment (shizhMSFT) [reply cmd/oras/root/push.go:203]: Your original code looks better to me
```go
mediaType := oras.MediaTypeUnknownConfig
if opts.Flag == option.ImageSpecV1_0 && opts.artifactType != "" {
	mediaType = opts.artifactType
}
```

**Review by shizhMSFT — CHANGES_REQUESTED** [review]
In fact, I have a more fundamental question: How do I specify both `artifactType` and `config.mediaType`? It seems not possible for image-spec 1.1 in the current PR.

We need the ability to set `config.mediaType` to `application/vnd.oci.image.config.v1+json` when `--artifact-platform` is set in order to support [OCI volumes](https://kubernetes.io/blog/2024/08/16/kubernetes-1-31-image-volume-source/). /cc @FeynmanZhou

**Comment by TerryHowe** [comment]
> In fact, I have a more fundamental question: How do I specify both `artifactType` and `config.mediaType`? It seems not possible for image-spec 1.1 in the current PR.

This is out of the scope for this PR and I don't think that was possible before this PR.

File: cmd/oras/internal/option/platform.go
Comment (shizhMSFT) [inline cmd/oras/internal/option/platform.go:86]: Can you make this flag as `[Experimental]` or `[Preview]`?

File: cmd/oras/root/push.go
Comment (TerryHowe) [reply cmd/oras/root/push.go:203]: The original code was clearer at least.

**Comment by shizhMSFT** [comment]
> This is out of the scope for this PR and I don't think that was possible before this PR.

Can you mark the flag `--artifact-platform` as `[Experimental]` or `[Preview]` so that we can complete it in next few iterations / PRs?

File: cmd/oras/root/push.go
Comment (shizhMSFT) [reply cmd/oras/root/push.go:203]: Can you apply your original code above to this PR?

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Comment by qweeah** [comment]
> In fact, I have a more fundamental question: How do I specify both artifactType and config.mediaType? It seems not possible for image-spec 1.1 in the current PR.

Through `oras push --artifact-type $ARTIFACT_TYPE --config $CONFIG_PATH:$CONFIG_TYPE`. It's by design `--artifact-platform` cannot be used to customize any media type.

**Comment by qweeah** [comment]
> We need the ability to set config.mediaType to application/vnd.oci.image.config.v1+json when --artifact-platform is set in order to support [OCI volumes](https://kubernetes.io/blog/2024/08/16/kubernetes-1-31-image-volume-source/). /cc @FeynmanZhou

In that case `--image-spec v1.0 --artifact-type application/vnd.oci.image.config.v1+json --artifact-platform xxx/xxx`

**Comment by shizhMSFT** [comment]
> > We need the ability to set config.mediaType to application/vnd.oci.image.config.v1+json when --artifact-platform is set in order to support [OCI volumes](https://kubernetes.io/blog/2024/08/16/kubernetes-1-31-image-volume-source/). /cc @FeynmanZhou
> 
> In that case `--image-spec v1.0 --artifact-type application/vnd.oci.image.config.v1+json --artifact-platform xxx/xxx`

This workaround should work.

**Comment by qweeah** [comment]
> > > We need the ability to set config.mediaType to application/vnd.oci.image.config.v1+json when --artifact-platform is set in order to support [OCI volumes](https://kubernetes.io/blog/2024/08/16/kubernetes-1-31-image-volume-source/). /cc @FeynmanZhou
> > 
> > 
> > In that case `--image-spec v1.0 --artifact-type application/vnd.oci.image.config.v1+json --artifact-platform xxx/xxx`
> 
> This workaround should work.

It just came to me that if `application/vnd.oci.image.config.v1` is a critical scenario that `oras push --platform` need to support, why not make it default? This default media type switch will only happen when `--platform` being used and thus is not breaking.

**Comment by TerryHowe** [comment]
> This default media type switch will only happen when `--artifact-platform` being used and thus is not breaking.

I'm not 100% following here, maybe create an issue if you think there is something. If `--artifact-platform` is used, it creates a 1.1 image right now. Without specifying the image version, it creates a 1.0 imager

**Comment by qweeah** [comment]
@TerryHowe What I am thinking is to use `application/vnd.oci.image.config.v1` instead of `application/vnd.unknown.config.v1+json` as default config media type. 

Actually, it requires oras to extra work on baking the config blob to meet below spec: https://github.com/opencontainers/image-spec/blob/da92727e9c8761ec087890466b8756b755aefd37/manifest.md#L68-73

```
   When the `config.mediaType` is set to `application/vnd.oci.image.config.v1+json`, the following additional restrictions apply:

  - The array MUST have the base layer at index 0.
  - Subsequent layers MUST then follow in stack order (i.e. from `layers[0]` to `layers[len(layers)-1]`).
  - The final filesystem layout MUST match the result of [applying](layer.md#applying-changesets) the layers to an empty directory.
  - The [ownership, mode, and other attributes](layer.md#file-attributes) of the initial empty directory are unspecified
```

This work goes below the scope of oras so what I am thinking is not doable for now.

**Comment by Wwwsylvia** [comment]
Just realized that the config media type has to be `application/vnd.oci.image.config.v1`, otherwise the end-to-end workflow won't work. See #1517 for details.

**Pull request merged**
